### PR TITLE
Closure handlers

### PR DESCRIPTION
**Pull request by dbrgn**
Converted handlers to boxed closures.

This is a bit tricky and involves unstable features.

The advantage is that the behavior of a handler thread can now be defined in a simple function.

@rnestler do you want to take a look at this?

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,12 @@
 language: rust
 sudo: false
 rust:
-  - stable
-  - beta
+#  - stable
+#- beta
```
Comment (rnestler) [inline .travis.yml:5]: Why commenting? Just remove it.

File: .travis.yml [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
   - nightly
-matrix:
-  allow_failures:
-    - rust: nightly
+#matrix:
+#  allow_failures:
+#    - rust: nightly
```
Comment (rnestler) [inline .travis.yml:9]: Same here, just remove it.

**Comment by rnestler** [comment]
@dbrgn I had a look at it. Closures seem indeed nicer than trait objects in some use cases :+1: 
Did you take a look at [FnBox](https://doc.rust-lang.org/std/boxed/trait.FnBox.html)? May be useful.

**Comment by rnestler** [comment]
Also a big :+1: for the commit messages :smile:

File: .travis.yml (hunk 2 above)
Comment (dbrgn) [inline .travis.yml:9]: It shows that it's only temporary. I hope to be able to use this on beta soon.

Should I still remove it?

**Comment by dbrgn** [comment]
> Did you take a look at FnBox? May be useful.

Yep, saw that. But it's still unstable, so I'll use my own typedef for now.

> Closures seem indeed nicer than trait objects in some use cases

Actually, unter der Dusche geriet ich ins Sinnieren. I'm not really sure why I'm registering a closure within a struct in the first place, as that struct doesn't really hold any information that we couldn't also pass directly to the function.

So in the end, might this also work with regular function invocations, without all this closure stuff? :joy: That would bring us back to compatibility with stable :)

> Also a big :+1: for the commit messages :smile: 

I had to rebase in two cases where I did it wrong again :grin:

**Comment by rnestler** [comment]
> I'm not really sure why I'm registering a closure within a struct in the first place, as that struct doesn't really hold any information that we couldn't also pass directly to the function.

Yes holding the closures in a struct doesn't make much sense

> So in the end, might this also work with regular function invocations, without all this closure stuff? :joy: That would bring us back to compatibility with stable :)

If you call the function like [that](https://github.com/coredump-ch/telegram-groups-bot/pull/1/files#diff-639fbc4ef05b315af92b4d836c31b023R93)

``` Rust
 94  +                        let handler_func: commands::BoxedHandler = match &*cmd.name {  
 95  +                            "help"   => Box::new(commands::handle_help),  
 96  +                            "groups" => Box::new(commands::handle_groups),  
 97  +                            _        => Box::new(commands::handle_debug),  
 98  +                        };
 99  +                        let handler = CommandHandler::new(&cmd, handler_func);
...
 104  +                            if let Some(msg) = handler.handle() {  
```

There is no point in using trait objects or closures, but if you want to create a generic list of command handlers that you can register it may be useful. Similar to the modifiers or sensor handlers in [spaceapi-server-rs](https://github.com/coredump-ch/spaceapi-server-rs).

**Comment by dbrgn** [comment]
Yeah. The initial idea was to mimick Iron's handlers. But while that gives you a lot of flexibility, it also adds a lot of overhead for simple handlers (the majority).

**Comment by dbrgn** [comment]
Replaced by #2.

**Pull request closed without merging**